### PR TITLE
Explore/jax container rocm

### PR DESCRIPTION
**Pull request by juliusroeder**

### CODE REVIEW

File: Lumi Jax container with cotainr/README.md [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+
+### Running Jax test suit
+
+The following command clones the Jax repository and runs a numpy test.
+
+```commandline
+ cd /project/$PROJECT_NUM/$USER/
+ mkdir jax 
+ cd jax
+ git clone https://github.com/jax-ml/jax.git
```
Comment (joasode) [inline Lumi Jax container with cotainr/README.md:111]: This would pull the main branch which would be significantly different to the Jax version installed through pip. Looking at the diff `lax_numpy_test.py` certainly has changed a bit https://github.com/jax-ml/jax/compare/jax-v0.5.0...main could this explain some of the tests failing?

File: Lumi Jax container with cotainr/conda_envs/install_jax.sh [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+# forward all gold linker requests to ld.lld. Symlinking the linker doesnt work
+touch /usr/bin/ld.gold
+echo -e '#!/bin/bash\n /opt/rocm-6.0.2/llvm/bin/ld.lld "$@"' > /usr/bin/ld.gold
+chmod +x /usr/bin/ld.gold
+
+source /opt/cotainr/conda/etc/profile.d/conda.sh
+conda activate conda_container_env
+git clone https://github.com/jax-ml/jax
+cd jax
+git fetch origin tag jax-v0.6.2
```
Comment (joasode) [inline Lumi Jax container with cotainr/conda_envs/install_jax.sh]: Do you mean to `git checkout jax-v0.6.2`? git fetch doesn't apply any changes to the local repo.

File: Lumi Jax container with cotainr/conda_envs/conda_env.yml [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    - torchaudio==2.3.1+rocm6.0
+    - torchvision==0.18.1+rocm6.0
+    - jax[rocm]
+    - absl-py
+    - cloudpickle
+    - filelock
+    - flatbuffers
+    - hypothesis
+    - mpmath>=1.3
+    - pillow>=10.4.0
```
Comment (joasode) [inline Lumi Jax container with cotainr/conda_envs/conda_env.yml:30]: There are conflicting packages, `pillow=10.2.0` and `pillow>=10.4.0`, as well as the mpmath. Is this neccessary?

File: Lumi Jax container with cotainr/conda_envs/conda_env.yml [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    - portpicker
+    - pytest-xdist
+    - rich
+    - matplotlib
+    - auditwheel
+variables:
+  LLVM_PATH: /opt/rocm/llvm
```
Comment (joasode) [inline Lumi Jax container with cotainr/conda_envs/conda_env.yml]: Question: Did this environment variable actually work for you? When I tried to compile CuPy, it was kindly ignored when building with cotainr. I had to manually export the variables before running cotainr to have it working.

**Review by joasode — CHANGES_REQUESTED** [review]
I think this BitBulldozer is small enough that the text is sufficient. But I have a few questions that may change some of the outcome.

File: Lumi Jax container with cotainr/README.md (hunk 1 above)
Comment (juliusroeder) [reply Lumi Jax container with cotainr/README.md:111]: yes! thanks! thats fixed it.

File: Lumi Jax container with cotainr/conda_envs/conda_env.yml (hunk 3 above)
Comment (juliusroeder) [reply Lumi Jax container with cotainr/conda_envs/conda_env.yml:30]: removed

File: Lumi Jax container with cotainr/conda_envs/conda_env.yml (hunk 4 above)
Comment (juliusroeder) [reply Lumi Jax container with cotainr/conda_envs/conda_env.yml]: These sometimes work

File: Lumi Jax container with cotainr/conda_envs/install_jax.sh (hunk 2 above)
Comment (juliusroeder) [reply Lumi Jax container with cotainr/conda_envs/install_jax.sh]: yea forgot to change it to 'git switch'. changed now

File: Lumi Jax container with cotainr/README.md [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+git switch --detach jax-v0.5.0
+```
+
+```commandline
+ srun --output=results.out --error=results.err --account=$PROJECT_NUM --time=00:15:00 --mem=120G --nodes=1 --ntasks-per-node=1 --cpus-per-task=16 --partition=dev-g --mpi=pmi2 --gpus-per-node=2 singularity exec -B /project/$PROJECT_NUM/$USER/jax/jax jax_torch_rocm6.2.4.sif  bash -c 'cd jax; python tests/lax_numpy_test.py --test_targets="testPad"'
+```
+
+Tests should all pass.
+
+### Running Jax example
```
Comment (joasode) [inline Lumi Jax container with cotainr/README.md:135]: This chapter is difficult to follow after the recent changes. If you want to preserve these errors in some way, I suggest to move them to the very bottom in a "## Error Archive" section where you can write something like "This and this version mismatch result in the following errors for the mnist classifier in the open source container", and move the errors there since they are not the key point of the Bulldozer anymore.

File: Lumi Jax container with cotainr/README.md [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+
+Lastly, we train a Mnist classifier using Jax. 
+
+In the Jax Github repository you first need to change line 33 in `examples/mnist_classifier.py` from `from examples import datasets` to `import datasets`.
```
Comment (joasode) [inline Lumi Jax container with cotainr/README.md:139]: Is this still required? Shouldn't this section be run with equivalent `git switch` as the previous section?

File: .gitignore [hunk 7]
```diff
@@ -160,3 +160,5 @@ cython_debug/
 #  and can be added to the global gitignore or merged into this file.  For a more nuclear
 #  option (not recommended) you can uncomment the following to ignore the entire idea folder.
 #.idea/
+
+.idea/
```
Comment (joasode) [inline .gitignore:162]: This is pretty funny combined with the (not recommended) comment above :)

File: Lumi Jax container with cotainr/README.md [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+Second, we run a small subset of the tests in the Jax Github repository.
+Last, we run the Jax Mnist classifier example from the Jax Github repository. 
+
+### Simple test
+
+This is a simple test to check if the GPUs are available.
+
+#### Open Source Container
+
+**Note: You may have to set the LD_LIBRARY_PATH**
```
Comment (joasode) [inline Lumi Jax container with cotainr/README.md:66]: What may we have to set LD_LIBRARY_PATH to, and why? I have a similar question for the note further down...

**Review by joasode — CHANGES_REQUESTED** [review]
I am happy that this turned out to positive results. Now there are just a few points of clarification and one section that is hard to follow.

File: Lumi Jax container with cotainr/README.md (hunk 6 above)
Comment (juliusroeder) [reply Lumi Jax container with cotainr/README.md:139]: Yes this is still required because the module stuff doesnt get picked up properly.

File: .gitignore (hunk 7 above)
Comment (juliusroeder) [reply .gitignore:162]: well sure its kinda not recommended where everyone in the team uses jetbrains but.... 
Anyways i just removed the commented out one.

File: Lumi Jax container with cotainr/README.md (hunk 8 above)
Comment (juliusroeder) [reply Lumi Jax container with cotainr/README.md:66]: done

**Review by joasode — APPROVED** [review]
I think it looks good now. Although you mentioned that you had also installed from pure conda and that worked as well from unofficial repositories. Do you want to include these experiments in this repo?
